File: common/src/main/java/com/ssblur/scriptor/word/subject/TouchSubject.java
package com.ssblur.scriptor.word.subject;

import com.ssblur.scriptor.helpers.targetable.EntityTargetable;
import com.ssblur.scriptor.helpers.targetable.LecternTargetable;
import com.ssblur.scriptor.helpers.targetable.Targetable;
import com.ssblur.scriptor.events.network.TraceNetwork;
import com.ssblur.scriptor.word.Spell;
import net.minecraft.world.entity.LivingEntity;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.phys.AABB;
import net.minecraft.world.phys.Vec3;

import java.util.List;
import java.util.concurrent.CompletableFuture;

public class TouchSubject extends Subject{

  @Override
  public CompletableFuture<List<Targetable>> getTargets(Targetable caster, Spell spell) {
    var result = new CompletableFuture<List<Targetable>>();
    if(caster instanceof EntityTargetable entityTargetable && entityTargetable.getTargetEntity() instanceof Player player) {
      TraceNetwork.requestTraceData(player, target -> result.complete(List.of(target)));
    } else if(caster instanceof LecternTargetable) {
      var pos = caster.getTargetBlockPos().relative(caster.getFacing());
      var entities = caster.getLevel().getEntitiesOfClass(
        LivingEntity.class,
        AABB.ofSize(
          new Vec3(
            pos.getX(),
            pos.getY(),
            pos.getZ()
          ),
          1,
          1,
          1
        )
      );

      if(entities.isEmpty())
        result.complete(List.of(new Targetable(caster.getLevel(), pos)));
      else
        result.complete(entities.stream().map(living -> (Targetable) new EntityTargetable(living)).toList());
    } else {
      result.complete(List.of(caster.simpleCopy()));
    }
    return result;
  }

  @Override
  public Cost cost() { return new Cost(1, COSTTYPE.ADDITIVE); }
}

File: common/src/main/kotlin/com/ssblur/scriptor/word/subject/TouchSubject.kt
package com.ssblur.scriptor.word.subject

import com.ssblur.scriptor.api.word.Subject
import com.ssblur.scriptor.helpers.targetable.EntityTargetable
import com.ssblur.scriptor.helpers.targetable.LecternTargetable
import com.ssblur.scriptor.helpers.targetable.Targetable
import com.ssblur.scriptor.network.server.TraceNetwork
import com.ssblur.scriptor.word.Spell
import com.ssblur.scriptor.word.descriptor.target.CollideWithWaterDescriptor
import net.minecraft.world.entity.LivingEntity
import net.minecraft.world.entity.player.Player
import net.minecraft.world.phys.AABB
import net.minecraft.world.phys.Vec3
import java.util.concurrent.CompletableFuture

class TouchSubject: Subject() {
  override fun getTargets(caster: Targetable, spell: Spell): CompletableFuture<List<Targetable>> {
    val result = CompletableFuture<List<Targetable>>()
    if (caster is EntityTargetable && caster.targetEntity is Player) {
      TraceNetwork.requestTraceData(
        caster.targetEntity as Player,
        spell.deduplicatedDescriptorsForSubjects().contains(CollideWithWaterDescriptor)
      ) { target: Targetable -> result.complete(listOf(target)) }
    } else if (caster is LecternTargetable) {
      val pos = caster.targetBlockPos.relative(caster.facing)
      val entities = caster.level.getEntitiesOfClass(
        LivingEntity::class.java,
        AABB.ofSize(
          Vec3(
            pos.x.toDouble(),
            pos.y.toDouble(),
            pos.z.toDouble()
          ),
          1.0,
          1.0,
          1.0
        )
      )

      if (entities.isEmpty()) result.complete(listOf(Targetable(caster.level, pos)))
      else result.complete(entities.stream().map { EntityTargetable(it) }.toList())
    } else {
      result.complete(java.util.List.of(caster.simpleCopy()))
    }
    return result
  }

  override fun cost() = Cost(1.0, COSTTYPE.ADDITIVE)
}
